teps/0094-configuring-resources-at-runtime.md

```diff
@@ -52,12 +61,495 @@ While this may be a valuable feature, it should be considered in a separate prop
 depending on the image or source being built.
 - Kubeflow pipelines and other data pipelines may have variable resource requirements
 depending on the data being processed.
-- Catalog Tasks should be generally reusable in different environments
+- Catalog `Task`s should be generally reusable in different environments
 that may have different resource constraints.
 
 ## Requirements
 
 - Users can specify `Step` and `Sidecar` resource requirements at runtime.
 - Users can specify `Step` and `Sidecar` resource requirements for `Task`s 
 or `Pipeline`s they don't own, especially those in the Catalog.
-- Users can specify resource requirements for individual `Step`s and `Sidecar`s.
\ No newline at end of file
+- Users can specify resource requirements for individual `Step`s and `Sidecar`s.
+
+## Proposal
+
+Augment `TaskRunSpec` and `PipelineTaskRunSpec` with a mapping of `Step` names to overrides
+and a mapping of `Sidecar` names to overrides.
+
+## Design Details
+
+```go
+import corev1 "k8s.io/api/core/v1"
+
+type TaskRunSpec struct {
+   ...
+   // A map of Step names to ResourceRequirements for that Step.
+   // A single entry with key "*" will apply the ResourceRequirements to all Steps.
+   // If ResourceRequirements are specified in both a Step and StepResources,
+   // the value from the StepResources will be used.
+   StepResources map[string]corev1.ResourceRequirements
```

Comment:
Great points! I didn't realize there was an applicable API convention here, and I agree we should use that.
The use of named sub-objects also addresses the issue of whether to use `Resources` alone or whether to use a Tekton-owned struct that allows future field overrides.

Good point about "*" opening us up to potential requests for supporting regex syntax, I'm not sure we would want to support that. To clarify, in the `applyTo` solution you're suggesting, this would be either a Step or Sidecar name or the special-case strings "allSteps" or "allSidecars"? It feels strange to me to allow the field to represent either a step name or an enum-like field; what if there is overlap? I've updated my proposal to specify that if the name is unspecified, it will apply to all Steps/Sidecars, but I recognize that this doesn't allow for future extensibility, and I'm open to better solutions.

Re: combining overrides for Steps and Sidecars: I would prefer to keep them separate since Steps and Sidecars use different structs. What if we want to override Step.Timeout? Or what if a Step and Sidecar share a name?

Reply:
> To clarify, in the applyTo solution you're suggesting, this would be either a Step or Sidecar name or the special-case strings "allSteps" or "allSidecars"?

could be either, but:

> It feels strange to me to allow the field to represent either a step name or an enum-like field; what if there is overlap? 

I agree it's weird to overload the field, so I was thinking two separate fields, which you wouldn't use together (even though that's a bit gross, it seems less bad than overloading one field maybe?)

For example something like this (also not attached to "applyTo" as the field name haha - or any of the names really):

```yaml
spec:
  stepResources:
  -  applyTo: allSteps/allSidecars
      name: my-step # it would be invalid to use both name and applyTo, but one of the two must be present
```

more ideas for name of "name" field: stepName, applyToStep

>  I've updated my proposal to specify that if the name is unspecified, it will apply to all Steps/Sidecars, but I recognize that this doesn't allow for future extensibility, and I'm open to better solutions.

Oh that's a nice option too!!! If we went that route, I think it actually could still allow for future extensibility b/c we could actually add a new field later if we need to without much trouble, e.g. say that for now this means apply to all steps and sidecars:

```yaml
spec:
  taskRef:
    name: image-build-task
  stepResources:
  -  requests:
       memory: 1Gi
```

But later if we want to support various options: "apply to all sidecars", "apply to all steps", "apply to all steps and sidecars" we could add a new field (and again you'd only ever use this field OR the name field)

```yaml
spec:
  taskRef:
    name: image-build-task
  stepResources:
  -  requests:
     memory: 1Gi
     applyTo: allStepsAndSidecars # this is our new field, and allStepsAndSidecars is the default value - so if neither name nor applyTo is explicitly present, it's as if you set applyTo: allStepsAndSidecars
```

>  I would prefer to keep them separate since Steps and Sidecars use different structs.

hmmm interesting - on the one hand this feels like an implementation detail to me - i.e. most users wont know that these are backed by different structs, but they CAN see that they have different fields so that's user facing 🤔   I was thinking of this feature as being more about the underlying container execution, which they both share. Timeout is an interesting example since sidecars dont have a timeout (it ALMOST makes sense that they would, but i can see why they wouldn't?). The only other example currently of their differences is `OnError` which I wouldn't expect to be able to override at runtime.

>  Or what if a Step and Sidecar share a name?

good point that might make my preference for combining the sections moot XD  uggggggggggggggggggggggggggggggh XD

i thought maybe this wouldnt be possible but I didnt realize that [we use different prefixes](https://github.com/tektoncd/pipeline/blob/28f950700e99fd22a175eab7e2c803248675cca0/pkg/pod/entrypoint.go#L54-L55)

i guess this means we do probably want 2 different sections - one more probably awful idea to throw out there so we can use one section, use `applyTo` (or similar) to indicate if it's a step or a sidecar? e.g. for a specific step:

```yaml
spec:
  stepResources:
  -  applyTo: step
      name: my-step
```

For a specific sidecar:

```yaml
spec:
  stepResources:
  -  applyTo: sidecar
      name: my-sidecar
```

Anyway probably not worth it but wanted to throw it out there XD

Reply:
> > I've updated my proposal to specify that if the name is unspecified, it will apply to all Steps/Sidecars, but I recognize that this doesn't allow for future extensibility, and I'm open to better solutions.
> 
> Oh that's a nice option too!!! If we went that route, I think it actually could still allow for future extensibility b/c we could actually add a new field later if we need to without much trouble.

Good point, I didn't think of that! If you're OK with this option, I think we should stick with the two options (apply to individual steps and apply to all), and add the `applyTo` field later if there is user demand. I'm anticipating some bikeshedding around whatever new syntax would be introduced for this field, and I'd prefer to start with the simpler option of an optional Step name and iterate.

> > I would prefer to keep them separate since Steps and Sidecars use different structs.
> 
> hmmm interesting - on the one hand this feels like an implementation detail to me - i.e. most users wont know that these are backed by different structs, but they CAN see that they have different fields so that's user facing 🤔 I was thinking of this feature as being more about the underlying container execution, which they both share. Timeout is an interesting example since sidecars dont have a timeout (it ALMOST makes sense that they would, but i can see why they wouldn't?). The only other example currently of their differences is `OnError` which I wouldn't expect to be able to override at runtime.

If we're introducing a way for TaskRuns to override Steps/Sidecars at runtime, is there a reason to distinguish between Container fields and other fields of Step and Sidecar? (Maybe yes, if we're assuming that any fields we add to Step and Sidecar should be decided only at author time?) It does seem to me like Steps and Sidecars are different from a user perspective.
 
> 
> > Or what if a Step and Sidecar share a name?
> 
> i guess this means we do probably want 2 different sections - one more probably awful idea to throw out there so we can use one section, use `applyTo` (or similar) to indicate if it's a step or a sidecar? e.g. for a specific step:
> 
> ```yaml
> spec:
>   stepResources:
>   -  applyTo: step
>       name: my-step
> ```
> 
> For a specific sidecar:
> 
> ```yaml
> spec:
>   stepResources:
>   -  applyTo: sidecar
>       name: my-sidecar
> ```
> 
> Anyway probably not worth it but wanted to throw it out there XD

I'd prefer to have separate structs rather than introducing a new syntax like this. Do you have a strong reason to prefer a single struct for both Step and Sidecar? If we want it to look as similar as possible from a user perspective, I can change the "taskRunStepName"/"taskRunSidecarName" to just "name".

Reply:
Resolving based on discussion from today's WG to keep Steps and Sidecars separate, and to allow only overriding named Steps and Sidecars.